pkg/channel/event_receiver.go

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2018 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package channel
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	cloudevents "github.com/cloudevents/sdk-go"
+	"go.uber.org/zap"
+	"knative.dev/eventing/pkg/kncloudevents"
+	"knative.dev/eventing/pkg/utils"
+)
+
+var (
+	shutdownTimeout = 1 * time.Minute
+
+	// ErrUnknownChannel is returned when an event is received by a channel dispatcher for a
+	// channel that does not exist.
+	ErrUnknownChannel = errors.New("unknown channel")
+)
+
+// EventReceiver starts a server to receive new events for the channel dispatcher. The new
+// event is emitted via the receiver function.
+type EventReceiver struct {
+	ceClient          cloudevents.Client
+	receiverFunc      ReceiverFunc
+	logger            *zap.Logger
+	hostToChannelFunc ResolveChannelFromHostFunc
+}
+
+// ReceiverFunc is the function to be called for handling the event.
+type ReceiverFunc func(context.Context, ChannelReference, cloudevents.Event) error
+
+// ReceiverOptions provides functional options to EventReceiver function.
+type ReceiverOptions func(*EventReceiver) error
+
+// ResolveChannelFromHostFunc function enables EventReceiver to get the Channel Reference from incoming request HostHeader
+// before calling receiverFunc.
+type ResolveChannelFromHostFunc func(string) (ChannelReference, error)
```

Comment:
Ideally we would change this to take in a URL, rather than just the host string. That will allow us to use the path to distinguish between Channels too.

Reply:
added a TODO in the code